### PR TITLE
perf(k3): implement iris barrier free lamport all reduce for small M

### PR DESCRIPTION
**Pull request by efric**
## Summary

This patch implements one shot all reduce with sentinel synchronization (i.e, barrier free lamport all reduce) for use in Kimi K3 AMD post MoE all reduce for M <= 6. 

The data itself supplies the readiness signal. Before publishing values to peer mailboxes, each GPU converts -0.0 to +0.0, leaving all other values unchanged. The mailboxes (receiver buffer) initially contain -0.0 sentinels, so receivers can poll incoming packs until no sentinel remains. The same loads provide both the data and its readiness indication, eliminating the separate cross-GPU ready-flag barrier. Three rotating mailbox generations protect data from premature reuse across calls, eliminating the separate completion barrier as well.

Microbenchmarks in us:

```
|  M |  Main | Candidate |  Gain |
| -: | ----: | --------: | ----: |
|  1 | 6.528 |     4.435 | 32.1% |
|  2 | 7.036 |     5.030 | 28.5% |
|  3 | 7.653 |     5.638 | 26.3% |
|  4 | 8.501 |     6.310 | 25.8% |
|  5 | 8.995 |     7.160 | 20.4% |
|  6 | 8.987 |     7.774 | 13.5% |
```

On TP8/EP8: results are effectively neutral or slight negative, so gated just for TP8/TP8;
Kimi K3 4K/1K, 8× MI350X, no eagle3, no prefill graph, median of three: 

```
   Topology      M    Main tok/s    Lamport tok/s    Change    Main TPOT    Lamport TPOT    Change

   TP8/TP8       1         55.56            57.19    +2.95%        17.53           17.00    −3.02%
  ──────────  ─────  ────────────  ───────────────  ────────  ───────────  ──────────────  ────────
                 2         71.89            74.38    +3.46%        27.00           26.17    −3.07%
  ──────────  ─────  ────────────  ───────────────  ────────  ───────────  ──────────────  ────────
                 4        135.33           138.13    +2.07%        28.10           27.55    −1.96%
  ──────────  ─────  ────────────  ───────────────  ────────  ───────────  ──────────────  ────────
                 6        187.28           189.71    +1.30%        30.05           29.63    −1.40%


```

## Test Plan

- Rebased onto `main` at `688bbb04`.
- 73 runtime tests passed for K3 communication arming, backend routing, and logits processing.
- 92 Iris host/compiler tests passed, including generated assembly checks for all eight gfx950 rank specializations.
- `pre-commit run --all-files` passed.
- The Lamport device code is unchanged by the rebase. The performance results above are prior measurements; multi-GPU correctness and performance were not rerun for this rebase.

### CODE REVIEW

**Review by nirvedhmeshram — APPROVED** [review]
LGTM
